File: src/conditions.py
from src.data import DataBank, clean_url
from src.parser import parse_descriptions, parse_image_url


def __get_condition_disease_status(type: str, data: DataBank):
    results: list[dict] = []

    for entry in data.get(type, []):
        result = dict()
        result["name"] = entry["name"]
        result["source"] = entry["source"]
        result["url"] = clean_url(
            f"https://5e.tools/conditionsdiseases.html#{entry['name']}_{entry['source']}"
        )

        result["description"] = []
        descriptions = parse_descriptions(
            "Description", entry["entries"], result["url"]
        )
        for name, text in descriptions:
            result["description"].append({"name": name, "text": text})

        result["image"] = None
        for fluff in data.get(f"{type}Fluff", []):
            if fluff["name"] == entry["name"] and fluff["source"] == entry["source"]:
                if "images" in fluff:
                    result["image"] = parse_image_url(fluff["images"])

        results.append(result)

    return results


def get_conditions_json(data: DataBank) -> list[dict]:
    results: list[dict] = []

    results.extend(__get_condition_disease_status("condition", data))
    results.extend(__get_condition_disease_status("status", data))

    return results


def get_diseases_json(data: DataBank) -> list[dict]:
    return __get_condition_disease_status("disease", data)

File: src/conditions.ts
import { Description, parseDescriptions, parseImageUrl } from './parser';
import { getConditionsDiseasesUrl } from './urls';

// Note, statuses and diseases also follow the same structure as Condition
interface Condition {
    name: string;
    source: string;
    url: string;
    description: Description[];
    image: string | null;
}

function getConditions(type: string, data: any): Condition[] {
    const results: Condition[] = [];

    const entries = data[type] || [];
    const fluffs = data[`${type}Fluff`] || [];
    for (const entry of entries) {
        const url = getConditionsDiseasesUrl(entry.name, entry.source);
        const result: Condition = {
            name: entry.name,
            source: entry.source,
            url: url,
            description: parseDescriptions('Description', entry.entries),
            image: null,
        };

        for (const fluff of fluffs) {
            if (fluff.name == entry.name && fluff.source == entry.source) {
                if (fluff.images) {
                    result.image = parseImageUrl(fluff.images);
                }
            }
        }

        results.push(result);
    }

    return results;
}

export function getConditionsStatusesAndDiseases(data: any): {
    conditions: Condition[];
    diseases: Condition[];
} {
    const conditions: Condition[] = [];
    conditions.push(...getConditions('condition', data));
    conditions.push(...getConditions('status', data));
    const diseases = getConditions('disease', data);

    return { conditions, diseases };
}
